File: src/assembler/grpc-docs-assembler.ts
/**
 * GrpcDocsAssembler — generates docs/api/grpc-reference.md
 * from the _TEMPLATE-GRPC-REFERENCE.md Nunjucks template.
 *
 * Interface-aware: only generates when the project config
 * includes a `grpc` interface. Returns empty array otherwise.
 *
 * Graceful no-op: if the source template does not exist in
 * resourcesDir, returns an empty array (backward compatibility).
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import { hasInterface } from "./conditions.js";

const TEMPLATE_PATH = "templates/_TEMPLATE-GRPC-REFERENCE.md";
const OUTPUT_SUBDIR = "api";
const OUTPUT_FILENAME = "grpc-reference.md";

/** Assembles docs/api/grpc-reference.md for gRPC-enabled projects. */
export class GrpcDocsAssembler {
  /** Generate gRPC API reference documentation. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): string[] {
    if (!hasInterface(config, "grpc")) {
      return [];
    }
    const templateFile = path.join(resourcesDir, TEMPLATE_PATH);
    if (!fs.existsSync(templateFile)) {
      return [];
    }
    const rendered = engine.renderTemplate(TEMPLATE_PATH);
    const destDir = path.join(outputDir, OUTPUT_SUBDIR);
    fs.mkdirSync(destDir, { recursive: true });
    const destFile = path.join(destDir, OUTPUT_FILENAME);
    fs.writeFileSync(destFile, rendered, "utf-8");
    return [destFile];
  }
}

File: java/src/main/java/dev/iadev/application/assembler/GrpcDocsAssembler.java
package dev.iadev.application.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/**
 * Assembles {@code contracts/api/grpc-reference.md} from the
 * {@code _TEMPLATE-GRPC-REFERENCE.md} Pebble template.
 *
 * <p>This is the fifteenth assembler in the pipeline
 * (position 15 of 23 per RULE-005). It is
 * <strong>interface-aware</strong>: only generates output
 * when the project configuration includes a {@code grpc}
 * interface type.</p>
 *
 * <p>Double graceful no-op:
 * <ol>
 *   <li>If no gRPC interface is configured, returns an
 *       empty list immediately (no I/O).</li>
 *   <li>If gRPC is configured but the template file does
 *       not exist, returns an empty list (backward
 *       compatibility).</li>
 * </ol>
 *
 * <p>Example usage:
 * <pre>{@code
 * Assembler grpcDocs = new GrpcDocsAssembler();
 * List<String> files = grpcDocs.assemble(
 *     config, engine, outputDir);
 * }</pre>
 * </p>
 *
 * @see Assembler
 * @see ConditionEvaluator#hasInterface
 * @see TemplateEngine#render(String, Map)
 */
public final class GrpcDocsAssembler implements Assembler {

    private static final String TEMPLATE_PATH =
            "shared/templates/_TEMPLATE-GRPC-REFERENCE.md";
    private static final String OUTPUT_SUBDIR = "contracts/api";
    private static final String OUTPUT_FILENAME =
            "grpc-reference.md";

    private final Path resourcesDir;

    /**
     * Creates a GrpcDocsAssembler using classpath resources.
     */
    public GrpcDocsAssembler() {
        this(resolveClasspathResources());
    }

    /**
     * Creates a GrpcDocsAssembler with an explicit resources
     * directory.
     *
     * @param resourcesDir the base resources directory
     */
    public GrpcDocsAssembler(Path resourcesDir) {
        this.resourcesDir = resourcesDir;
    }

    /**
     * {@inheritDoc}
     *
     * <p>Checks for gRPC interface presence first, then
     * verifies template existence before rendering with
     * Pebble and writing to
     * {@code api/grpc-reference.md}.</p>
     */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        if (!ConditionEvaluator.hasInterface(
                config, "grpc")) {
            return List.of();
        }
        Path templateFile =
                resourcesDir.resolve(TEMPLATE_PATH);
        if (!Files.exists(templateFile)) {
            return List.of();
        }
        Map<String, Object> context =
                ContextBuilder.buildContext(config);
        String rendered = engine.render(
                TEMPLATE_PATH, context);
        Path destDir =
                outputDir.resolve(OUTPUT_SUBDIR);
        CopyHelpers.ensureDirectory(destDir);
        Path destFile =
                destDir.resolve(OUTPUT_FILENAME);
        CopyHelpers.writeFile(destFile, rendered);
        return List.of(destFile.toString());
    }

    private static Path resolveClasspathResources() {
        return dev.iadev.util.ResourceResolver
                .resolveResourcesRoot(TEMPLATE_PATH, 3);
    }
}
